### PR TITLE
Enable ThreadSanitizer on macOS

### PR DESCRIPTION
**Pull request by pavelkumbrasev**
Signed-off-by: pavelkumbrasev <pavel.kumbrasev@intel.com>

### Description 
To enable ThreadSanitizer on __APPLE__ (basically on macOS) replaced  semaphore_t on dispatch_semaphore_t which supports via ThreadSanitizer.
In accordance to scenarios where and how we use semaphore the performance should not change, but still need to prove this.

Fixes # - #358

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [x] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
@phprus

### CODE REVIEW

**Comment by akukanov** [comment]
Does this change introduce any new dependencies for either compilation/link time or execution time which need to be documented?

Do we know that the dispatch semaphore has no spurious wakeups?

**Comment by pavelkumbrasev** [comment]
> Does this change introduce any new dependencies for either compilation/link time or execution time which need to be documented?
> 
> Do we know that the dispatch semaphore has no spurious wakeups?

I currently researching both questions. For now I checked that dispatch_semaphore_t introduced with macOS 10.6. Also I did not observe any new flags during the compilation. (I compiled example from #358 using dispatch_semaphore_t without any additional flags)

**Comment by pavelkumbrasev** [comment]
It seems that from dispatch_semaphore_t [documentation](https://developer.apple.com/documentation/dispatch/1452919-dispatch_semaphore_signal) there should not be spurious wakeups during `wait` calls.  Also from the [implementation](https://opensource.apple.com/source/libdispatch/libdispatch-84.5/src/semaphore.c) point of view they have similar while loop on case DISPATCH_TIME_FOREVER.

File: test/tbb/test_partitioner.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
-    Copyright (c) 2021 Intel Corporation
+    Copyright (c) 2022 Intel Corporation
```
Comment (aleksei-fedotov) [inline test/tbb/test_partitioner.cpp:2]: Why replacing? Isn't it should be extended instead?
```suggestion
    Copyright (c) 2021-2022 Intel Corporation
```

Comment (omalyshe) [reply test/tbb/test_partitioner.cpp:2]: Agree with @aleksei-fedotov . While changing the existing file the year range should be used.

**Review by aleksei-fedotov — APPROVED** [review]
Looks good to me!

**Comment by natale-p** [comment]
Tested on mac os 12.6, and I can easily see that I don't have many false positives as I had without this patch.

**Comment by natale-p** [comment]
Hello, any news on this PR? Can it be merged?

**Comment by pavelkumbrasev** [comment]
Hi @natale-p, we are going to merge it after we will satisfy all test requirements for it. (I hope it will be soon)

**Pull request merged**